### DISCUSSION TITLE
LDR Components

### DISCUSSION THREAD

**Discussion by GrimmInsomnia** [dsharlet/LiveSPICE#168] [Ideas]
I am yet to get completely familiar with the code of LiveSpice, but was wanting to work with some LDRs in some circuits. 
I've been working on an LDR file for LiveSpice based off of the existing files, backwards engineering, and web help/tools. 
I haven't worked out adding the component and getting it fully functional, but I figured someone here may or may want to take this code, expand upon it and get it working and added!

        using ComputerAlgebra;
        using System;
        using System.ComponentModel;
        
        namespace Circuit
        {
            /// <summary>
            /// LightDependentResistor is a linear component with V = R(i)*i.
            /// </summary>
            [Category("Generic")]
            [DisplayName("LDR")]
            [DefaultProperty("Resistance")]
            [Description("Light-dependent resistor.")]
            public class LightDependentResistor : TwoTerminal
            {
                protected Quantity resistanceDark = new Quantity(10000, Units.Ohm);
                protected Quantity resistanceLight = new Quantity(100, Units.Ohm);
                protected double lightLevel = 0.5;
        
                [Serialize, Description("Dark resistance of this LDR.")]
                public Quantity ResistanceDark { get { return resistanceDark; } set { if (resistanceDark.Set(value)) NotifyChanged(nameof(ResistanceDark)); } }
        
                [Serialize, Description("Light resistance of this LDR.")]
                public Quantity ResistanceLight { get { return resistanceLight; } set { if (resistanceLight.Set(value)) NotifyChanged(nameof(ResistanceLight)); } }
        
                [Serialize, Description("Light level at which the resistance of this LDR is the average of its dark and light resistances.")]
                public double LightLevel { get { return lightLevel; } set { if (value >= 0 && value <= 1) { lightLevel = value; } } }
        
                public LightDependentResistor() { Name = "LDR1"; }
        
                public static Expression Analyze(Analysis Mna, string Name, Node Anode, Node Cathode, Expression LightLevel, Expression ResistanceDark, Expression ResistanceLight)
                {
                    // i = V/R(light)
                    Expression i = (Anode.V - Cathode.V) / (ResistanceLight * LightLevel + ResistanceDark * (1 - LightLevel));
                    Mna.AddPassiveComponent(Anode, Cathode, i);
                    return i;
                }
        
                public static Expression Analyze(Analysis Mna, Node Anode, Node Cathode, Expression LightLevel, Expression ResistanceDark, Expression ResistanceLight)
                {
                    return Analyze(Mna, "", Anode, Cathode, LightLevel, ResistanceDark, ResistanceLight);
                }
        
                public override void Analyze(Analysis Mna) { Analyze(Mna, Name, Anode, Cathode, lightLevel, resistanceDark, resistanceLight); }
        
                public static void Draw(SymbolLayout Sym, double x, double y1, double y2, int N, double Scale)
                {
                    double h = y2 - y1;
        
                    Sym.DrawFunction(
                        EdgeType.Black,
                        (t) => x - Scale * (Math.Abs((t + 0.5) % 2 - 1) * 2 - 1),
                        (t) => t * h / N + y1,
                        0, N, N * 2);
                }
        
                public static void Draw(SymbolLayout Sym, double x, double y1, double y2, int N) { Draw(Sym, x, y1, y2, N, (y2 - y1) / (N + 1)); }
        
                protected internal override void LayoutSymbol(SymbolLayout Sym)
                {
                    base.LayoutSymbol(Sym);
        
                    Sym.AddWire(Anode, new Coord(0, 16));
                    Sym.AddWire(Cathode, new Coord(0, -16));
                    Sym.InBounds(new Coord(-10, 0), new Coord(10, 0));
        
                    Draw(Sym, 0, -16, 16, 7);
                }
            }
        }

The formatting may be a bit off from the copy and paste but this is the first section of code.

**Comment by dsharlet** [comment]
Your code looks quite close to working, I can't actually spot any problems just based on what you've posted. What isn't working when you try to use this? The only real issue I see is that the only "light source" we have is going to be some arbitrary expression you can type into an input box (we aren't simulating any lights/LEDs...).

A few minor things you could do in the code:
- You could use `Resistor.Analyze` to handle the analysis after you've computed the resistance, like this: https://github.com/dsharlet/LiveSPICE/blob/master/Circuit/Components/VariableResistor.cs#L85
- You might want `LightLevel` to be an `Expression` rather than a `double`. That way, people could use expressions that are more interesting than just constants (functions of time, or grabbing the value of other quantities in the simulation).

All that said though, this seems to me to be basically equivalent to `VariableResistor`, except replacing the wiper position with light level. Unless we have some special treatment for lighting in the simulation to make use of this parameter, I am not sure it makes sense to add this (but I think it is a great thing for you to experiment with if you want to learn more about LiveSPICE!).

**Reply by GrimmInsomnia** [reply]
Thanks for your awesome response!
I'm yet to compile this code into my LiveSPICE download but will do so and update accordingly. 

I'm not sure why I didn't think to use the structure of the `VariableResistor`, great idea! Tying in the simulation of LED's to base the resistance level of the LDR was my ultimate plan of course but I am yet to do so properly. With your idea, I'll make the adjustments and share them for input!

I very much love this application and its utilitarian capabilities. I was actually programming one myself (not in C#, however) until I came across LiveSPICE! Though it has functioned properly for the most part thus far, I hope to expand upon the audio tracking and simulation for some of my current audio circuit projects. That being said, I am sure to dive deeper!

Thanks for your input. Maybe this can be added and expanded upon for future updates!
